pcap/search.go

```diff
@@ -136,22 +136,62 @@ func genICMPFilter(src, dst net.IP) PacketFilter {
 	}
 }
 
-// XXX currently assumes legacy pcap is produced by the input reader
 // XXX need to handle searching over multiple pcap files
 func (s *Search) Run(ctx context.Context, w io.Writer, r pcapio.Reader) error {
+	_, err := s.Reader(r).WriteToWithContext(ctx, w)
+	return err
+}
+
+type SearchReader struct {
+	*Search
+	reader pcapio.Reader
+	opts   gopacket.DecodeOptions
+	npkt   int
+}
+
+func (s *Search) Reader(r pcapio.Reader) *SearchReader {
 	opts := gopacket.DecodeOptions{Lazy: true, NoCopy: true}
-	var npkt int
+	return &SearchReader{Search: s, reader: r, opts: opts}
+}
+
+func (s *SearchReader) WriteToWithContext(ctx context.Context, w io.Writer) (n int64, err error) {
 	for {
 		if err := ctx.Err(); err != nil {
+			return n, err
+		}
+		block, err := s.next()
+		if err != nil {
+			return n, err
+		}
+		if block == nil {
+			return n, nil
+		}
+		if err := writeFull(w, block); err != nil {
+			return n, err
+		}
```

Comment:
Nit: This isn't strictly necessary since the [io.Writer documentation](https://pkg.go.dev/io?tab=doc#Writer) says you'll get a non-nil error from a short Write.

Reply:
Oh great. Thanks @nwt. Great to get rid of this.